libsession/src/main/java/org/session/libsession/snode/SnodeAPI.kt

```diff
@@ -598,398 +486,269 @@ object SnodeAPI {
-            return parseEnvelopes(newRawMessages)
-        } else {
-            listOf()
-        }
-    }
+    fun parseRawMessagesResponse(rawResponse: RawResponse, snode: Snode, publicKey: String, namespace: Int = 0, updateLatestHash: Boolean = true, updateStoredHashes: Boolean = true): List<Pair<SignalServiceProtos.Envelope, String?>> =
+        (rawResponse["messages"] as? List<*>)?.let { messages ->
+            if (updateLatestHash) updateLastMessageHashValueIfPossible(snode, publicKey, messages, namespace)
+            removeDuplicates(publicKey, messages, namespace, updateStoredHashes).let(::parseEnvelopes)
+        } ?: listOf()
 
     fun updateLastMessageHashValueIfPossible(snode: Snode, publicKey: String, rawMessages: List<*>, namespace: Int) {
         val lastMessageAsJSON = rawMessages.lastOrNull() as? Map<*, *>
         val hashValue = lastMessageAsJSON?.get("hash") as? String
-        if (hashValue != null) {
-            database.setLastMessageHashValue(snode, publicKey, hashValue, namespace)
-        } else if (rawMessages.isNotEmpty()) {
-            Log.d("Loki", "Failed to update last message hash value from: ${rawMessages.prettifiedDescription()}.")
+        when {
+            hashValue != null -> database.setLastMessageHashValue(snode, publicKey, hashValue, namespace)
+            rawMessages.isNotEmpty() -> Log.d("Loki", "Failed to update last message hash value from: ${rawMessages.prettifiedDescription()}.")
         }
     }
 
     fun removeDuplicates(publicKey: String, rawMessages: List<*>, namespace: Int, updateStoredHashes: Boolean): List<*> {
-        val originalMessageHashValues = database.getReceivedMessageHashValues(publicKey, namespace)?.toMutableSet() ?: mutableSetOf()
+        val originalMessageHashValues = database.getReceivedMessageHashValues(publicKey, namespace) ?: emptySet()
         val receivedMessageHashValues = originalMessageHashValues.toMutableSet()
-        val result = rawMessages.filter { rawMessage ->
-            val rawMessageAsJSON = rawMessage as? Map<*, *>
-            val hashValue = rawMessageAsJSON?.get("hash") as? String
-            if (hashValue != null) {
-                val isDuplicate = receivedMessageHashValues.contains(hashValue)
-                receivedMessageHashValues.add(hashValue)
-                !isDuplicate
-            } else {
-                Log.d("Loki", "Missing hash value for message: ${rawMessage?.prettifiedDescription()}.")
-                false
+        return rawMessages.filter { rawMessage ->
+            (rawMessage as? Map<*, *>)
+                ?.let { it["hash"] as? String }
+                ?.let { receivedMessageHashValues.add(it) }
+                ?: false.also { Log.d("Loki", "Missing hash value for message: ${rawMessage?.prettifiedDescription()}.") }
+        }.also {
+            if (updateStoredHashes && originalMessageHashValues.containsAll(receivedMessageHashValues)) {
```

Comment:
It seems you have inverted the logic here by checking `containsAll` instead of `!=`

Reply:
fixed! 🎉